### DISCUSSION TITLE
Ordering by fields omitted for inputs

### DISCUSSION THREAD

**Discussion by jagged3dge** [MichalLytek/typegraphql-prisma#237] [Q&A]
Hi all!

I'm in a quandary. I have a model with basic audit log fields `createdAt`, `updatedAt` & `deletedAt`. 
I do not want users to be able to input their values for these fields, hence I've defined these like so:

```
model Brand {
  id        Int             @id @default(autoincrement())
  name      String          @unique

  /// @TypeGraphQL.omit(input: true)
  createdAt DateTime?       @default(now())

  /// @TypeGraphQL.omit(input: true)
  updatedAt DateTime?       @default(now()) @updatedAt

  /// @TypeGraphQL.omit(input: true)
  deletedAt DateTime?

 // ... other relation fields with User and Campaigns
}
```

This is my GraphQL query:

```gql
query GetBrands($where: BrandWhereInput, $orderBy: [BrandOrderByWithRelationInput!], $take: Int, $skip: Int) {
  brands(where: $where, orderBy: $orderBy, take: $take, skip: $skip) {
    id
    name
    createdAt
    updatedAt
    deletedAt
  }
}
```

If I want to retrieve brands in order of their `updatedAt` date in descending order, I cannot use the following variables with the above query: 
```ts
{
  orderBy: { 
    updatedAt: 'desc'
  }
}
```

This query throws the following error:

```
Variable "$orderBy" got invalid value { updatedAt: "desc" }; Field "updatedAt" is not defined by type "BrandOrderByWithRelationInput".
```

I could write a custom resolver class that extends `BrandOrderByWithRelationInput` to resolve this. 

But I have plenty of models that will require this. Manually creating custom resolver classes for all the models is tedious and a maintenance nightmare. 

Is there a simpler workaround?

Edit:

After careful thought, I see why the generator doesn't create the fields in the `XXXOrderByWithRelationInput` because they are marked to be omitted in all input args. 

However, considering that the `XXXOrderByWithRelationInput` class (and other `XXXOrderBy...` classes) is/are only used in queries, never in mutations, I propose that there must be some way to indicate that these fields should be allowed for ordering by.

Just like `input: boolean` and `output: boolean`, there could be an option: `order: boolean` that would be checked only if a field is marked with `input: false`. 

If `input: false, order: true`, generator would enable the field in the `XXXOrderBy...Input` class(es). 
If `input: false, order: false` (or only `input: false`), then the current process continues as is.

**Comment by jagged3dge** [comment]
For now, I'm able to work around this by doing so:

1. Create a Prisma middleware that prevents updating the audit-log fields during any of create/createMany/update/updateMany queries
2. Remove `/// @TypeGraphQL.omit(input: true)` from both `createdAt` and `updatedAt` fields in necessary models in `prisma.schema`
3. Re-run the generator

**Comment by MichalLytek** [comment]
Have you tried the extended input omit feature?
```
/// @TypeGraphQL.omit(input: ["update", "where", "orderBy"])
```

**Reply by jagged3dge** [reply]
Not yet. 
I'll recreate a minimal repro from that earlier project with its settings and see how it goes with the extended input omit feature.
